BUG: allow empty multiindex (fixes .isin regression, GH16777)

drudd opened the pull request:
- [X] closes issue #16777
 - [X] tests added / passed
 - [X] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [X] whatsnew entry

Commit:
BUG: allow empty multiindex (fixes .isin regression, GH16777)

Commit:
BUG: allow empty multiindex (fixes .isin regression, GH16777)

jorisvandenbossche commented on pandas/core/indexes/multi.py:
```diff
@@ -89,6 +89,9 @@ def __new__(cls, levels=None, labels=None, sortorder=None, names=None,
             raise ValueError('Length of levels and labels must be the same.')
+        if names is not None and len(names) > 0 and len(levels) == 0:
+            levels = [[] for _ in names]
+            labels = [[] for _ in names]
```
I suppose this is to fix `from_tuples` to be able to work with empty list of tuples? If so, I would rather do this there instead of here (I mean, making sure `__init__` gets passed those empty lists of levels and labels)

jorisvandenbossche replied:
or to be more precise (after looking at the code), make sure that `from_arrays` gets a list of empty lists of the correct length

jreback replied:
there was a recently merged pr where we had this discussion 

iirc was sym_diff for empty input

drudd replied:
I added the code here to ensure all creation paths had the same behavior. Otherwise the same checks and empty array creation code would need to be added to both `.from_tuples` and `.from_array`.

drudd replied:
@jreback is this the PR you're thinking of: PR #16486? It looks like that change to `_shallow_copy_with_infer` could be refactored to simply ensuring metadata was propagated through to `_shallow_copy`.

jreback replied:
yes, I believe you can remove those changes (and just let it pass thru here then)

jorisvandenbossche commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -766,6 +766,14 @@ def test_from_arrays_empty(self):
             MultiIndex.from_arrays(arrays=[])
 
+        # 0 levels, names defined
+        result = MultiIndex.from_arrays(arrays=[],
```
So my suggestion is that here you would need to pass `arrays=[[], []]`

drudd replied:
Right, which would keep the current behavior for `.from_arrays`. I'd argue that if `.from_tuples(tuples=(), names=names)` is valid then `.from_array(arrays=[], names=names)` should also be allowed (and produce an identical object).

jreback replied:
I think ``arrays=[]`` is reasonable here (can you add a test for both ``[[], []]`` and ``[]`` (for ``.from_arrays``)

jorisvandenbossche replied:
The reasoning for requiring `[[], []]` in `from_arrays` is that `[]` has the wrong shape for the input, as opposed to in the case of `from_tuples`, where `[]` is the correct shape of what it expects as input.

But OK, it is a bit subjective, so I am fine with either.

drudd replied:
I like the shape argument for maintaining the original behavior of arrays, so I'm refactoring to only allow empty lists for `from_tuples`. Happy to revisit this later.

drudd replied:
Actually I did find one reason to keep the empty multi-index checks in the `__new__` function. If I remove this code (or run on the current master branch), then these two ways of constructing an empty multi-index lead to non-equivalent indexes, in the sense of failing `assert_index_equal`:

```
In [1]: import pandas as pd

In [2]: A = pd.MultiIndex(levels=[[], []], labels=[[], []], names=['a', 'b'])

In [3]: B = pd.MultiIndex.from_arrays(arrays=[[], []], names=['a', 'b'])

In [4]: A
Out[4]:
MultiIndex(levels=[[], []],
           labels=[[], []],
           names=[u'a', u'b'])

In [5]: B
Out[5]:
MultiIndex(levels=[[], []],
           labels=[[], []],
           names=[u'a', u'b'])

In [6]: pd.testing.assert_index_equal(A, B)
...
AssertionError: MultiIndex level [0] are different

MultiIndex level [0] classes are not equivalent
[left]:  Index([], dtype='object', name=u'a')
[right]: Float64Index([], dtype='float64', name=u'a')
```

drudd replied:
The above is due to conversions that happen in `_factorize_from_iterables` that do not occur in `__new__`. I think it makes sense for an empty multi-index to not have base type, and the code I originally wrote in `__new__` ignores the indexes created in `from_arrays`. If someone wanted to create an empty multi-index with typed levels, however, this could break that.

jreback replied:
so ``.from_arrays`` will always factorize and hence the ``Float64Index`` result. This is a separate issue (and we need to do this for 0.21 as its technically an API change, pls open an issue). We can do the empty check on all levels  to avoid the factorization (as its proper to have these be ``object`` dtype and not float).

jreback replied:
going back to the original, I think @jorisvandenbossche spelling is right. ``arrays=[]`` is not valid (``arrays=None`` would be ok though if names were passed)

jorisvandenbossche reviewed (commented):
Nice PR! added some a suggestion that *might* make it less hacky in the `__init__`

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -2618,8 +2620,9 @@ def _wrap_joined_index(self, joined, other):
-            return algos.isin(self.values,
-                              MultiIndex.from_tuples(values).values)
+            values = MultiIndex.from_tuples(values,
+                                            names=self._levels).values
```
this is odd to pass _levels as names

drudd replied:
Great point, that obviously should be `_names`.

Commit:
Replace incorrect use of self._levels with self._names

jreback commented on doc/source/whatsnew/v0.20.3.txt:
```diff
@@ -51,6 +51,7 @@ Indexing
 ^^^^^^^^
 
 - Bug in ``Float64Index`` causing an empty array instead of ``None`` to be returned from ``.get(np.nan)`` on a Series whose index did not contain any ``NaN`` s (:issue:`8569`)
+- Bug in ``MultiIndex`` causing an error when passing an empty iterable to `.isin`. (:issue:`16777`)
```
say ``MultiIndex.isin`` instead

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -2618,8 +2620,9 @@ def _wrap_joined_index(self, joined, other):
         if level is None:
-            return algos.isin(self.values,
-                              MultiIndex.from_tuples(values).values)
+            values = MultiIndex.from_tuples(values,
```
you *might* be able to change this to ``_ensure_index(values)`` I think (as its more idiomatic). I also don't think you will need the ``.values``

jorisvandenbossche replied:
`_ensure_index` is not meant to support values for a MultiIndex I think:

```
In [159]: pd.core.indexes.base._ensure_index([(1, 2), (3, 4)])
Out[159]: Index([(1, 2), (3, 4)], dtype='object')
```

drudd replied:
Not able to address this before 0.20.3

jreback commented on pandas/tests/indexes/test_multi.py:
```diff
@@ -1717,6 +1725,12 @@ def test_from_tuples(self):
                                'from empty list',
                                MultiIndex.from_tuples, [])
 
+        result = MultiIndex.from_tuples([], names=['a', 'b'])
```
similary I *think* ``[(), ()]`` should work here

jorisvandenbossche replied:
No, IMO this should not work here, as in this case you actually do pass 2 values, but empty values. So IMO this should raise.

jreback replied:
I agree with this. It should raise because you have a mismatch between the names and what you passed. If you had NOT passed an empty list it would be ok.

jreback replied:
can you make the empty case a separately named test

drudd replied:
Done.

TomAugspurger commented:
@drudd are you able to update this in the next few days? Releasing 0.20.3 by the end of the week. Are there any questions on the implementation or desired behavior?

drudd commented:
Hi Tom, sorry, I was traveling over the weekend. I should be able to clean this up hopefully address comments today.

Commit:
Requested change to whats-new

Commit:
Use names property rather than _names in MultiIndex.isin

jreback commented:
@drudd if you can fix up. we want to include with 0.20.3, releasing tomorrow.

Commit:
Revert ability to construct MultiIndex from empty arrays, ensure all tests pass despite issue #16844

drudd commented:
Okay, I reported the issue of factorization (#16844), which unfortunately is difficult to work around. I've added some checks in the `.from_arrays` function to prevent `.from_arrays(arrays=[])` from being valid and adjusted tests accordingly. Not as clean as I'd like but any further changes are api breaking.

@jreback I'm not sure I followed your last comment on the `MultiIndex.from_tuples([], names=['a', 'b'])` test. You agree with the test but think it should raise? My fix to the `.isin` issue relies on `.from_tuples` succeeding when passed an empty list, which unlike `.from_arrays` I believe to be correct behavior as it's the right input shape.

Commit:
Remove empty list checks in MultiIndex.__new__

Commit:
Remove empty array check in .from_arrays, falling back to check in __new__

jreback commented on pandas/core/indexes/multi.py:
```diff
@@ -1129,10 +1129,11 @@ def from_tuples(cls, tuples, sortorder=None, names=None):
+            if names is None:
+                msg = 'Cannot infer number of levels from empty list'
+                raise TypeError(msg)
+            arrays = [[]]*len(names)
```
this is causing a lint issue

drudd replied:
Fixed.

jreback reviewed (changes requested):
couple of issues

Commit:
Fix for pep8

Commit:
Split empty from_tuples test

jreback committed:
add commit number to test

jreback commented:
thanks!
